sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala

```diff
@@ -178,44 +176,75 @@ class StatePartitionReader(
     }
   }
 
+  private def processListStateEntries(stateVarName: String): Iterator[InternalRow] = {
+    if (partition.sourceOptions.flattenCollectionTypes) {
+      store
+        .iterator(stateVarName)
+        .flatMap { pair =>
+          val key = pair.key
+          val result = store.valuesIterator(key, stateVarName)
+          result.map { entry =>
+            SchemaUtil.unifyStateRowPair((key, entry), partition.partition)
+          }
+        }
+    } else {
+      store
+        .iterator(stateVarName)
+        .map { pair =>
+          val key = pair.key
+          val result = store.valuesIterator(key, stateVarName)
+          var unsafeRowArr: Seq[UnsafeRow] = Seq.empty
+          result.foreach { entry =>
+            unsafeRowArr = unsafeRowArr :+ entry.copy()
+          }
```

Comment:
Have you considered using a mutable collection for unsafeRowArr to get rid of the overhead of copying? As far as I understand, there's no risk of race conditions here, and you also convert it to an immutable array before you pass it to another function, so it can't be unexpectedly modified outside. Wdyt?

Reply:
Done